### PR TITLE
Report author-provided accessible labels in Chrome that are not visible elsewhere on the page

### LINKED ISSUE #4773: Report label in browse mode if it isn't rendered anywhere else (e.g. aria-label)
**Reported by khsruru on 2015-01-05 07:56**
Hello, I’m Hyongsop Kim in Korea.
I have a suggestion about NVDA and aria-label.
Let’s suppose that one html code is blow.
`<input type=”text” aria-label=”type keyword” </input>`
In that case, if I press tab to approach to the edit box, NVDA reads aria-label attribute.
But in browse mode, if I press arrow key to approach, NVDA reads only edit, not read the aria-label text in Internet Explorer and Firefox.
In my opinion, aria-label text is used for the screen reader users in case not available to write visual text, NVDA should read the aria-label text even if use arrow keys.
Please check this issue.

Thank you.

### PR DESCRIPTION
**Pull request by michaelDCurran**
This is a follow-on from pr #7513. This was originally fixed for Firefox but not Chrome, as chrome did not yet implement the necessary APIs. Now it does, this standardizes the code to use a technique supported by both browsers.

### Link to issue number:
Fixes #4773

### Summary of the issue:
Sometimes a web author will use aria-label on a node to convey screen-reader specific information, however the node also has useful and valid content itself (E.g. an edit field, or an iFrame).
NVDA already reports this label when moving to these controls with quick navigation or tabbing, but it should also report it when reading the control with the browse mode caret, as this label will not be rendered anywhere else.
Of course care must be taken to ensure this is in deed a label set by aria-label and not say a label from aria-labelledby or an associated html label tag.  The one exception to this is if the associated label node is hidden from the screen.
 
### Description of how this pull request fixes the issue:
In Chrome we now convey to NVDA that these labels must always be reported, by exposing the 'alwaysReportName' attribute on these nodes from the vbufbackend.

### Testing performed:
Tested Firefox and Chrome browsers by arrowing up and down the page with NVDA's screen layout mode off, using the document from 
[sr_only_labels.html.txt](https://github.com/nvaccess/nvda/files/1240739/sr_only_labels.html.txt)
Note: the needed functionality was introduced to Chrome in version 66.
 
### Known issues with pull request:
This is now implemented for Firefox and Chrome, but 
it is not supported in other browsers because:
* Edge has no means of asking a node if its name (label) has been explicitly set by the author
* We are not expending any effort for Internet Explorer now, though PRs will be accepted from the community 

### Change log entry:
Bug fixes:
Accessible labels for controls in Google Chrome are now more readily reported in browse mode when the label does not appear as content itself. (#4773)

### CODE REVIEW

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 1]
```diff
@@ -552,14 +558,13 @@ VBufStorage_fieldNode_t* GeckoVBufBackend_t::fillVBuf(IAccessible2* pacc,
 		|| role == ROLE_SYSTEM_BUTTONMENU
 		|| ((role == ROLE_SYSTEM_CHECKBUTTON || role == ROLE_SYSTEM_RADIOBUTTON) && !isLabelVisible(pacc));
 	// Whether this node has a visible label somewhere else in the tree
-	const bool labelVisible = canDetectLabelVisibility // Not all browsers support getting a node's labelledBy node
-		&& nameIsExplicit && name && name[0] //this node must actually have an explicit name, and not be just an empty string
+	const bool labelVisible = nameIsExplicit && name && name[0] //this node must actually have an explicit name, and not be just an empty string
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: Do we care what character `name[0]` is? If it is required to be visible, do we need to test that it is not a "non-printable" character?

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 2]
```diff
@@ -552,14 +558,13 @@ VBufStorage_fieldNode_t* GeckoVBufBackend_t::fillVBuf(IAccessible2* pacc,
...
-	const bool labelVisible = canDetectLabelVisibility // Not all browsers support getting a node's labelledBy node
-		&& nameIsExplicit && name && name[0] //this node must actually have an explicit name, and not be just an empty string
+	const bool labelVisible = nameIsExplicit && name && name[0] //this node must actually have an explicit name, and not be just an empty string
 		&&(!nameIsContent||role==ROLE_SYSTEM_TABLE) // We only need to know if the name won't be used as content or if it is a table (for table summary)
 		&&isLabelVisible(pacc); // actually do the check
 	// If the node is explicitly labeled for accessibility, and we haven't used the label as the node's content, and the label does not visibly appear anywhere else in the tree (E.g. aria-label on an edit field)
 	// then ensure that the label is always reported along withe the node
 	// We must exclude tables from this though as table summaries / captions are handled very specifically
-	if(canDetectLabelVisibility&&nameIsExplicit&&!nameIsContent&&(role!=ROLE_SYSTEM_TABLE)&&!labelVisible) {
+	if(nameIsExplicit&&!nameIsContent&&(role!=ROLE_SYSTEM_TABLE)&&!labelVisible) {
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: I understand this is consistent with the style of the rest of the file, however I would find this much easier to read with spaces around the boolean operators.

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 3]
```diff
@@ -552,14 +558,13 @@ VBufStorage_fieldNode_t* GeckoVBufBackend_t::fillVBuf(IAccessible2* pacc,
 		|| role == ROLE_SYSTEM_BUTTONMENU
 		|| ((role == ROLE_SYSTEM_CHECKBUTTON || role == ROLE_SYSTEM_RADIOBUTTON) && !isLabelVisible(pacc));
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: `isLabelVisible()` is called here and on line 563. I wonder if this is intentional, should the bool `labelVisible` which is created at line 561 be used instead?

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp (hunk 1 above)
Comment (feerrenrut) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: I think this question is still valid, even though github has hidden it.

Comment (michaelDCurran) [reply nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: I believe what we have is correct, as  if nameIsExplicit, then the author has provided some kind of accessible label, and if it contains anything (even a space) clearly they are wanting to inforce the content, E.g. overriding existing descendant content with a space.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**